pkg/ddl/partition.go

```diff
@@ -3869,54 +3837,65 @@ func (w *reorgPartitionWorker) BackfillData(handleRange reorgBackfillTask) (task
+			// we must check if old IDs already been written,
+			// i.e. double written by StateWriteOnly or StateWriteReorganization.
+			// TODO: while waiting for BatchGet to check for duplicate, do another round of reads in parallel?
+			found, err = txn.BatchGet(ctx, w.oldKeys)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+
+		for i, prr := range w.rowRecords {
+			taskCtx.scanCount++
+			key := prr.key
+
 			if len(w.oldKeys) > 0 {
-				// we must check if old IDs already been written,
-				// i.e. double written by StateWriteOnly or StateWriteReorganization.
-				// The good thing is that we can then also skip the index generation for that row and we don't need to
-				// check if duplicate index entries was already copied either!
-				// TODO: while waiting for BatchGet to check for duplicate, do another round of reads in parallel?
-				found, err = txn.BatchGet(ctx, w.oldKeys)
+				if _, ok := found[string(w.oldKeys[i])]; ok {
+					// Already filled, i.e. double written earlier by concurrent DML
+					continue
+				}
+
+				// Check if we can write the old key, since there can still be concurrent update
+				// happening on the rows from fetchRowColVals(), we can involve them in this
+				// transaction and fail when committing, so we will retry with a new snapshot.
+				// TODO: Test a concurrent pessimistic transaction that reads/writes concurrently here,
+				// and commits after after this transaction.
+				err = txn.Set(w.oldKeys[i], prr.vals)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				// Ensure the key does not exist!
+				// It actually works without this assertion, but it may help troubleshooting.
+				err = txn.SetAssertion(w.oldKeys[i], kv.SetAssertNotExist)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				// Also don't actually write it :)
+				err = txn.Delete(w.oldKeys[i])
 				if err != nil {
 					return errors.Trace(err)
 				}
-			}
-
-			for i := 0; i < w.records; i++ {
-				taskCtx.scanCount++
 
-				if len(w.oldKeys) > 0 {
-					if _, ok := found[string(w.oldKeys[i])]; ok {
-						// Alredy filled
-						continue
-					}
-					tbl := w.reorgedTbl.GetPartition(w.newPids[i])
-					if tbl == nil {
-						return dbterror.ErrUnsupportedReorganizePartition.GenWithStackByArgs()
-					}
-					// TODO: is this looking up each index entry at a time or in an optimistic way and only checks
-					// at commit time?
-					// AddRecord will assign a new _tidb_rowid, since we don't provide one.
-					_, err = tbl.AddRecord(w.tblCtx, txn, w.rows[i])
-					if err != nil {
-						return errors.Trace(err)
-					}
-					taskCtx.addedCount++
+				// Due to EXCHANGE PARTITION, the existing _tidb_rowid may collide between partitions!
+				// Generate new _tidb_rowid.
+				recordID, err := tables.AllocHandle(w.ctx, w.tblCtx, w.reorgedTbl)
+				if err != nil {
+					return errors.Trace(err)
 				}
+
+				// tablecodec.prefixLen is not exported, but is just TableSplitKeyLen + 2
```

Comment:
Not get the comments point. Why `+2` here?

Reply:
TableSplitKeyLen is `t_<encoded tableID>` only, and we want to include the `r_` as well, so that is where the +2 comes from.

Reply:
Use `len(recordPrefixSep)` instead of 2?

Reply:
Hmm, all names that start with lower case is not exported :(
I can change those names, but then the PR grows a bit with unrelated changes...
Is it OK if I create a follow-up issue+PR for exporting RecordPrefixSepLength later?